File: POO/Constructores_Clases/Interface/GO/Reproducible.go
package main

import "fmt"

// Interfaz
type Reproducible interface {
    Reproducir() string
    Pausar() string
    Detener() string
    Info() string
}

// Implementación
type Audio struct {
    titulo  string
    artista string
    duracion int // en segundos
}

// Constructor sin parámetros
func NewAudio() *Audio {
    return &Audio{
        titulo:  "Sin título",
        artista: "Desconocido",
        duracion: 0,
    }
}

func (a *Audio) Reproducir() string {
    return fmt.Sprintf("Reproduciendo: %s - %s", a.titulo, a.artista)
}

func (a *Audio) Pausar() string {
    return "Audio pausado"
}

func (a *Audio) Detener() string {
    return "Audio detenido"
}

func (a *Audio) Info() string {
    minutos := a.duracion / 60
    segundos := a.duracion % 60
    return fmt.Sprintf("%s - %s (%d:%02d)", a.titulo, a.artista, minutos, segundos)
}

func main() {
    var reproducible Reproducible = NewAudio()
    fmt.Println(reproducible.Info())
    fmt.Println(reproducible.Reproducir())
}

File: POO/Constructores_Clases/Interface/JAVA/Reproducible.java
package Interface.JAVA;
public interface Reproducible {
    String reproducir();
    String pausar();
    String detener();
    String info();
}

class Video implements Reproducible {
    private String titulo;
    private int duracion;
    private boolean reproduciendo;

    // Constructor sin parámetros
    public Video() {
        this.titulo = "Video sin título";
        this.duracion = 0;
        this.reproduciendo = false;
    }

    @Override
    public String reproducir() {
        reproduciendo = true;
        return "Reproduciendo: " + titulo;
    }

    @Override
    public String pausar() {
        reproduciendo = false;
        return "Video pausado";
    }

    @Override
    public String detener() {
        reproduciendo = false;
        return "Video detenido";
    }

    @Override
    public String info() {
        String estado = reproduciendo ? "Reproduciendo" : "Detenido";
        return titulo + " - " + duracion + "s - " + estado;
    }

    public void mostrar() {
        System.out.println(info());
    }

    public static void main(String[] args) {
        Reproducible video = new Video();
        video.mostrar();
        System.out.println(video.reproducir());
        video.mostrar();
    }
}